### PR TITLE
BZ1839982 FOR 2.3: Updating access mode for LiveMigration

### PR DESCRIPTION
**Pull request by bgaydosrh**
This PR is for the 2.3 repo. 

Please label **peer review needed** and **enterprise-4.4**.

Test Build: http://file.bos.redhat.com/bgaydos/071320_2/cnv/cnv_live_migration/cnv-live-migration.html

Once content approved for this PR I will update https://github.com/openshift/openshift-docs/pull/22789 with same content for 2.4 repo.

Bob

### CODE REVIEW

File: modules/cnv-understanding-live-migration.adoc [hunk 1]
```diff
@@ -5,16 +5,10 @@
...
 = Understanding live migration
 
-Live migration is the process of moving a running virtual machine instance to 
-another node in the cluster without interruption to the virtual workload or 
-access. This can be a manual process, if you select a virtual machine instance 
-to migrate to another node, or an automatic process, if the 
-virtual machine instance has a `LiveMigrate` eviction strategy and the node on 
-which it is running is placed into maintenance. 
-
-[IMPORTANT]
```
Comment (aburdenthehand) [inline modules/cnv-understanding-live-migration.adoc:15]: I would keep this admonition in this module as it's still relevant to understanding live migration.

File: modules/cnv-understanding-live-migration.adoc [hunk 2]
```diff
@@ -5,16 +5,10 @@
 [id="cnv-understanding-live-migration_{context}"]
 = Understanding live migration
 
-Live migration is the process of moving a running virtual machine instance to 
-another node in the cluster without interruption to the virtual workload or
```
Comment (aburdenthehand) [inline modules/cnv-understanding-live-migration.adoc:9]: This might just be me being particular, but in general it would be better to edit text rather than block-copy the already existing text in a new PR. The reason being is that it means you can't use 'git blame' to easily figure out why something is written a certain way. This becomes important in subsequent releases when you need to follow the trail to a BZ or jira, and git blame is an easy way to see changes line-by-line and chase down the commit in which the text in question was introduced, which will have the ID. Replacing block text by copying existing content means git blame will cul-de-sac with this PR, which is unrelated.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// Module included in the following assemblies:
+//
+// * cnv/live_migration/cnv-live-migration.adoc
+
+[id="cnv-updating-access-mode-to-enable-live-migration_{context}"]
+= Updating access mode to enable LiveMigration
+
+Storage class defaults in {VirtProductName} are configured to provide ReadWriteOnce (RWO) access mode
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:8]: 2.3 docs should use the CNV product variable

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+by default. Therefore, LiveMigration is not enabled by default unless you
+alter the access mode of your virtual machine.
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:10]: The access mode is for the storage class, not the virtual machine (as I understand it)

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 5]
```diff
@@ -0,0 +1,30 @@
+// Module included in the following assemblies:
+//
+// * cnv/live_migration/cnv-live-migration.adoc
+
+[id="cnv-updating-access-mode-to-enable-live-migration_{context}"]
+= Updating access mode to enable LiveMigration
+
+Storage class defaults in {VirtProductName} are configured to provide ReadWriteOnce (RWO) access mode
+by default. Therefore, LiveMigration is not enabled by default unless you
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:9]: LiveMigration is enabled by default iirc. It just requires this accessmode to work.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ReadWriteMany (RWX) access mode to be live migrated.
+
+.Procedure
+To override the default RWO access mode, update the storage class configuration
+by doing the following:
+
+. Export the storage class object.
++
+----
+$ export STORAGE_CLASS="nfs"
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:23]: This command is not exporting the object, rather it is exporting (to the shell environment) the 'STORAGE_CLASS' variable. In the context of this procedure, this step is unnecessary in an example like this because it is simpler just to use the value, 'nfs', in the following 'oc patch' command.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+----
++
+. Update the storage class `kubevirt-storage-class-defaults` from RWO to RWX access mode.
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:26]: `kubevirt-storage-class-defaults` is the name of the ConfigMap

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+[id="cnv-updating-access-mode-to-enable-live-migration_{context}"]
+= Updating access mode to enable LiveMigration
+
+Storage class defaults in {VirtProductName} are configured to provide ReadWriteOnce (RWO) access mode
+by default. Therefore, LiveMigration is not enabled by default unless you
+alter the access mode of your virtual machine.
+
+.Prerequisites
+* In order to use LiveMigration, virtual machines must have a persistent volume claim (PVC) with a shared
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:13]: While this is required to live migrate a VM, I don't think it's a prerequisite for updating a ConfigMap

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 9]
```diff
@@ -0,0 +1,30 @@
+// Module included in the following assemblies:
+//
+// * cnv/live_migration/cnv-live-migration.adoc
+
+[id="cnv-updating-access-mode-to-enable-live-migration_{context}"]
+= Updating access mode to enable LiveMigration
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:6]: This procedure is really about editing the ConfigMap so that the storageclass uses the ReadWriteMany accessMode required for live migration to function properly. Since this procedure already requires a storageclass to have had its storage settings added to the ConfigMap, I would suggest it would be better for the procedure to focus on adding the correct storage settings in the first place. Or maybe we can add an admonition or Prerequisite to the live migration assembly that links to the [existing content](https://docs.openshift.com/container-platform/4.4/cnv/cnv_virtual_machines/cnv_virtual_disks/cnv-storage-defaults-for-datavolumes.html) about these storage settings?

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+. Export the storage class object.
++
+----
+$ export STORAGE_CLASS="nfs"
+----
++
+. Update the storage class `kubevirt-storage-class-defaults` from RWO to RWX access mode.
++
+----
+$ oc patch -n openshift-cnv cm kubevirt-storage-class-defaults -p '{"data":{"'$<STORAGE_CLASS>'.accessMode":"ReadWriteMany"}}'
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:29]: The previous step explicitly exported the 'STORAGE_CLASS' variable. You don't need to wrap it in '<' '>' because it's now a literal variable rather than a replaceable one. (this is more of a FYI since I think we should drop the previous command anyway)

File: modules/cnv-understanding-live-migration.adoc (hunk 2 above)
Comment (bgaydosrh) [reply modules/cnv-understanding-live-migration.adoc:9]: Gotcha, will just edit existing content in the future.

File: modules/cnv-understanding-live-migration.adoc (hunk 1 above)
Comment (bgaydosrh) [reply modules/cnv-understanding-live-migration.adoc:15]: Kept. I went back and forth on keeping the admonition as I made it a prereq in the other module, but whatever you think will make it more noticeable.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 9 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:6]: Moved the prereq to the assembly and added the xref in the new intro.

* Before using LiveMigration, ensure that the storage class used by the 
virtual machine has a persistent volume claim (PVC) with a shared 
ReadWriteMany (RWX) access mode. 
See xref:../../cnv/cnv_virtual_machines/cnv_virtual_disks/cnv-storage-defaults-for-datavolumes[Storage defaults for DataVolumes]
to ensure all your storage settings are correct.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 3 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:8]: Changed.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 5 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:9]: Gotcha. Changed to: 

For LiveMigration to function properly, the storage class must use the 
ReadWriteMany (RWX) access mode.

Storage class defaults in {CNVProductName} are configured to provide 
ReadWriteOnce (RWO) access mode by default. To enable LiveMigration, the 
access mode must be changed to RWX.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 4 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:10]: Thanks. Should be covered by above change.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 8 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:13]: I removed this prereq in favor of the prereq in the assembly with the cross-ref described above.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 6 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:23]: Removed the export step.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 7 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:26]: Changed to:

To override the default RWO access mode, update the *ConfigMap* 
`kubevirt-storage-class-defaults` storage class to use RWX 
access mode.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 10 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:29]: Noted. Makes sense now.

File: modules/cnv-understanding-live-migration.adoc [hunk 11]
```diff
@@ -5,16 +5,14 @@
...
-====
-Virtual machines must have a PersistentVolumeClaim (PVC) with a shared 
-ReadWriteMany (RWX) access mode to be live migrated.
-====
+This process is manual if you select a virtual machine instance
+to migrate to another node. The process is automatic if the
+virtual machine instance has a `LiveMigrate` eviction strategy and the node on
+which it is running is placed into maintenance.
 
+[IMPORTANT]
```
Comment (aburdenthehand) [inline modules/cnv-understanding-live-migration.adoc:16]: Needs the appropriate markup. Also "ReadWriteMany".
Might be easier to 'git checkout -- cnv-understanding-live-migration.adoc` if there are no changes to the module.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 9 above)
Comment (aburdenthehand) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:6]: This has not been addressed

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+//
+// * cnv/live_migration/cnv-live-migration.adoc
+
+[id="cnv-updating-access-mode-to-enable-live-migration_{context}"]
+= Updating access mode to enable LiveMigration
+
+For LiveMigration to function properly, your storage class must use the
+RWX access mode.
+
+Storage class defaults in {CNVProductName} are configured to provide
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:11]: This isn't entirely accurate. The default storage class uses a RWO accessMode by default, but subsequent storage classes are added to the ConfigMap by the user, in which case they specify the accessMode then.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+ReadWriteOnce (RWO) access mode by default. To enable LiveMigration, the
```
Comment (aburdenthehand) [inline modules/cnv-updating-access-mode-to-enable-live-migration.adoc:12]: Again, I don't think 'enable' is an appropriate word here. LiveMigration requires RWX for the storage class, but it doesn't 'enable' it. You can specify multiple storage classes, with differing settings, in the ConfigMap.

File: modules/cnv-understanding-live-migration.adoc (hunk 11 above)
Comment (bgaydosrh) [reply modules/cnv-understanding-live-migration.adoc:16]: Reverted this to original by making appropriate edits. Was not able to figure out how to remove the changed file from GitHub. Doing git checkout -- would have prior to commit but not sure how to revert if once it's in GitHub. if you have info, let me know.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 12 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:11]: _Changed to the following, dropping the paragraph beginning with "Storage class defaults...."_

Default storage classes use a ReadWriteOnce (RWO) access mode by default. 
When you add additional storage classes to the `ConfigMap`, specify the
appropriate access mode. 

For LiveMigration to function properly, your storage class must use the
ReadWriteMany (RWX) access mode.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 13 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:12]: I struck this paragraph entirely (see previous comment).

I am striking the word enable, so this module is being renamed _cnv-updating-access-mode-for-live-migration_.

File: modules/cnv-updating-access-mode-to-enable-live-migration.adoc (hunk 9 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-to-enable-live-migration.adoc:6]: From our Slack conversation of Monday June 22, it sounds like the problem here was (as below) the clarification of the task and it's reflection in the title of the module. I believe this to be clearer now that I have eliminated "enable" and changed the wording, but let's meet in some dynamic medium like Blue Jeans or Google Meet if there are still issues you feel should be addressed.

File: modules/cnv-updating-access-mode-for-live-migration.adoc [hunk 14]
```diff
@@ -0,0 +1,23 @@
+// Module included in the following assemblies:
+//
+// * cnv/live_migration/cnv-live-migration.adoc
+
+[id="cnv-updating-access-mode-for-live-migration_{context}"]
+= Updating access mode for LiveMigration
+
+Default storage classes use a ReadWriteOnce (RWO) access mode by default.
```
Comment (awels) [inline modules/cnv-updating-access-mode-for-live-migration.adoc:8]: Storage classes do NOT define RWO or RWX, storage classes define reclaimPolicy and volumeBinding modes. Persistent Volumes define which accessModes they support. If we have a dynamic provisioner that creates PVs, then the provisioners configuration will determine if it can create them. For manually created PVs the yaml/json of the PV will determine what access modes it supports.

The accessMode is one of the criteria kubernetes uses to determine if it can match a PVC with a PV. So if a PVC requests RWX, but no PVs with RWX are available it will not match them.

In short this has NOTHING to do with storage classes.

File: modules/cnv-updating-access-mode-for-live-migration.adoc [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+When you add additional storage classes to the *ConfigMap*, you specify the
+appropriate access mode.
+
+For LiveMigration to function properly, your storage class must use the
+ReadWriteMany (RWX) access mode. Use this 
+procedure to update the access mode, if needed.
+
+.Procedure
+* Update the *ConfigMap*
```
Comment (awels) [inline modules/cnv-updating-access-mode-for-live-migration.adoc:17]: I guess I understand why you thought this was related to storage classes, the configmap is keyed on storage class. But the config map is just for the UI to determine if the user selects a storage class which accessMode to create the PVC with.

So the configmap basically says. If you use this storage class in the UI, then the UI will know to create the matching DataVolume (and thus PVC) with these accessModes. But from a kubernetes perspective there is no relation between storage class and accessModes.

**Review by awels — COMMENTED** [review]
So the whole config map is a little confusing IMO. We are trying to make an easy statement where we say storage class = X accessMode. Which is not true in general. The configmap is just a hint for the UI to pick appopriate accessMode for the DV(PVC) it is creating based on the storage class the user selected.

Its just from a kubernetes perspective there is no relationship between storage class and access modes.

**Review by bgaydosrh — COMMENTED** [review]
In the wake of @awels last review, I have revised this content mostly from his input, taking the focus off of the config map and the discussion of storage class in general.

**Comment by awels** [comment]
looks fine to me.

**Review by adellape — COMMENTED** [review]
Same suggestions as noted in https://github.com/openshift/openshift-docs/pull/22789#pullrequestreview-455002156.

**Comment by bgaydosrh** [comment]
@adellape - these are changes you noted in https://github.com/openshift/openshift-docs/pull/22789

Ready for check and merge to enterprise-4.4 only.

File: modules/cnv-updating-access-mode-for-live-migration.adoc [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+[id="cnv-updating-access-mode-for-live-migration_{context}"]
+= Updating access mode for LiveMigration
+
+For LiveMigration to function properly, you must use the
+ReadWriteMany (RWX) access mode. Use this
+procedure to update the access mode, if needed.
+
+.Procedure
+* To set the RWX access mode, run the following `oc patch` command.
```
Comment (adellape) [inline modules/cnv-updating-access-mode-for-live-migration.adoc:13]: s/command./command:/

**Review by adellape — APPROVED** [review]
Same tiny nit from https://github.com/openshift/openshift-docs/pull/22789#pullrequestreview-456888870, but overall LGTM.

File: modules/cnv-updating-access-mode-for-live-migration.adoc (hunk 16 above)
Comment (bgaydosrh) [reply modules/cnv-updating-access-mode-for-live-migration.adoc]: Done
